http/http.go

```diff
@@ -18,9 +20,89 @@ import (
 
 var tracer = trace.GlobalTracer
 
-// shared code for POSTing to an endpoint, that consumes JSON, that is zlib-
+// HTTPClientTracer is a wrapper around a `net/http/httptrace` that handles
+// proper tracing and metric emission.
+type HTTPClientTracer struct {
+	prefix      string
+	stats       *statsd.Client
+	traceClient *trace.Client
+	mutex       *sync.Mutex
+	ctx         context.Context
+	currentSpan *trace.Span
+}
+
+// NewHTTPClientTrace makes a new HTTPClientTracer with new span created from
+// the context.
+func NewHTTPClientTrace(ctx context.Context, stats *statsd.Client, tc *trace.Client, prefix string) *HTTPClientTracer {
+	span, _ := trace.StartSpanFromContext(ctx, "http.start")
+	return &HTTPClientTracer{
+		prefix:      prefix,
+		stats:       stats,
+		traceClient: tc,
+		mutex:       &sync.Mutex{},
+		ctx:         ctx,
+		currentSpan: span,
+	}
+}
+
+// GetClientTrace is a convenience to return a filled-in `ClientTrace`.
+func (hct *HTTPClientTracer) GetClientTrace() *httptrace.ClientTrace {
+	return &httptrace.ClientTrace{
+		GotConn:              hct.GotConn,
+		DNSStart:             hct.DNSStart,
+		GotFirstResponseByte: hct.GotFirstResponseByte,
+		ConnectStart:         hct.ConnectStart,
+		WroteRequest:         hct.WroteRequest,
+	}
+}
+
+// StartSpan is a convenience that replaces the current span in our
+// tracer and flushes the outgoing one.
+func (hct *HTTPClientTracer) StartSpan(name string) *trace.Span {
+	hct.mutex.Lock()
+	defer hct.mutex.Unlock()
+	newSpan, _ := trace.StartSpanFromContext(hct.ctx, name)
+	hct.currentSpan.ClientFinish(hct.traceClient)
```

Comment:
...huh, I didn't think that this would result in reasonable output (since the tracer's callbacks can be invoked in any order and in parallel), but evidently it does. I wonder if we're lying to ourselves by pretending that these spans happen in sequence though...